packages/tizen_rpc_port/example/example_proxy/lib/main.dart

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 Samsung Electronics Co., Ltd. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'package:flutter/material.dart';
+
+import 'package:flutter/services.dart';
+import 'package:tizen_log/tizen_log.dart';
+
+import 'message_proxy.dart';
+
+// ignore_for_file: public_member_api_docs
+
+const String _logTag = 'TizenRpcPortProxyExample';
+
+void main() {
+  runApp(const MyApp());
+}
+
+class MyApp extends StatefulWidget {
+  const MyApp({super.key});
+
+  @override
+  State<MyApp> createState() => _MyAppState();
+}
+
+String _message = '';
```

Comment:
Thanks for this suggestion.
But, I'm sorry I think we cannot modify this implementation.

The Proxy application registers the generated delegate in the stub.
The stub invokes the delegate registered by the Proxy application when it's needed.
To invoke the delegation method(callback), we has to know the delegate id and the sequence number.
Proxy applications can remove or add registered delegates by themselves.

It's the intentional design.
To invoke the delegation method(callback), we has to know the delegate id and the sequence number.
The CallbackBase class is implemented by inheriting the Parcelable class. It's for sending and receiving the delegate id and the sequence number.

The TIDL interface is defined by the developer. The delegate method is not always one.
```
interface Message {
  void onRegisteredEvent(string name) delegate;
  void onReceivedEvent(string name, string message) delegate;
  ...
 
  int Register(string name, onRegisteredEvent registerEvent, onReceivedEvent receivedEvent);
  int Send(string name, string message);
}
```

Reply:
C# provided delegate interface.
Delegate method (such as `NotifyCallback`) can be multiple.
`NotifyCallback` can't be included in `Message`. These two are 'dependency' relationships.
The current implementation is the implementation we intended, and I don't think this design is wrong.
But if there is a better way, you can change it.
Your suggestion currently recommended is difficult to apply to us.

Reply:
First of all, assuming that you applied one of my above designs, what prevents you from declaring multiple delegate methods?

Reply:
I'm sorry. Correct the parts that can cause confusion.
the delegate method is an `aggregation` relationship, not a `dependency` relationship, because it is stored inside when registering.
However, the `inheritance` relationship is not appropriate because it can be replaced internally.

Reply:
Do you have a concrete example of why the design is not feasible? Notice that there's no inheritance relationship between `Message` and `NotifyCallback` in my designs.

Reply:
> First of all, assuming that you applied one of my above designs, what prevents you from declaring multiple delegate methods?

I'm sorry, but I didn't understand this.
Can you explain it in more detail?
Do you mean inheriting and implementing multiple interfaces?

Reply:
Our TIDL design intention should be that each delegate can be replaced at runtime. If you inherit and implement it, it is difficult to carry out that part. In this example, of course, inheritance implementation is possible. However, it is not appropriate for TIDL's intended direction.
In this example, there can be various classes that have inherited `NotifyCallback`.

Reply:
If the implementation is changed according to the first proposal, there is a problem that onReceived() can be declared duplicate.
If the implementation is changed according to the second proposal, the delete id and sequence number for the registered function must be stored and managed separately. And, the flag(once) has to be added to the registration method.

To make serialization easier, we create objects and deliver them.

Reply:
@hjhun Thank you for your answer!

> If the implementation is changed according to the first proposal, there is a problem that onReceived() can be declared duplicate.

Yes, you're right, but the code in the above is just an example and the actual implementation may vary depending on the user's design. I directly extended the `NotifyCallback` interface from the `Message` class in the above example just for easy sharing of the `message` variable between them. The user doesn't always have to "implement" the callback inside the proxy base class and can define a dedicated class just like `MyNotify` in the current code.

> If the implementation is changed according to the second proposal, the delete id and sequence number for the registered function must be stored and managed separately. And, the flag(once) has to be added to the registration method.

Actually the delegate ID, the sequence number, the once flag, and (de-)serialization are all implementation details which should be hidden from the user (app developer)'s point of view. Yes, according to my proposal, the information such as delegate ID and sequence numbers should be stored in another data structure as you told. Is it really impractical to change the design?

Reply:
Also can I see a C# equivalent of the generated code if you have?

Reply:
> @hjhun Thank you for your answer!
> 
> > If the implementation is changed according to the first proposal, there is a problem that onReceived() can be declared duplicate.
> 
> Yes, you're right, but the code in the above is just an example and the actual implementation may vary depending on the user's design. I directly extended the `NotifyCallback` interface from the `Message` class in the above example just for easy sharing of the `message` variable between them. The user doesn't always have to "implement" the callback inside the proxy base class and can define a dedicated class just like `MyNotify` in the current code.
> 
> > If the implementation is changed according to the second proposal, the delete id and sequence number for the registered function must be stored and managed separately. And, the flag(once) has to be added to the registration method.
> 
> Actually the delegate ID, the sequence number, the once flag, and (de-)serialization are all implementation details which should be hidden from the user (app developer)'s point of view. Yes, according to my proposal, the information such as delegate ID and sequence numbers should be stored in another data structure as you told. Is it really impractical to change the design?

It's not impractical. 
We understand what you means. However, it seems better to maintain similarity with the generated code (C or C++ or C#) with the existing TIDL and deliver the once flag in the object. If the object does not include the once flag, the developer must deliver the unintended once flag to the argument of the method.

In the current example, it seems possible to inherit and implement NotifyCallback as you said.

@upple,
Could you change the implementation to the suggested code?

Reply:
The following code is C# code by generated tidlc.

```
/*
 * Generated by tidlc 1.10.3.
 */

using System;
using System.Collections.Generic;
using System.Threading;
using Tizen.Applications;
using Tizen.Applications.RPCPort;
namespace RPCPort
{
namespace message_proxy
{
    namespace Proxy
    {
        public class Message : ProxyBase
        {
            public event EventHandler Connected;
            public event EventHandler Disconnected;
            public event EventHandler Rejected;

            private static readonly string _tidlVersion = "1.10.3";
            private bool _online = false;
            private string _appId;
            private Object _lock = new Object();
            private List<CallbackBase> _delegateList = new List<CallbackBase>();

            public abstract class CallbackBase
            {
                internal int Id;
                internal int SeqId;
                internal bool Once;
                private static volatile int _seqNum = 0;

                public string Tag
                {
                    get
                    {
                        return Id.ToString() + "::" + SeqId.ToString();
                    }
                }

                public CallbackBase(int delegateId, bool once)
                {
                    Id = delegateId;
                    SeqId = _seqNum++;
                    Once = once;
                }

                internal virtual void OnReceivedEvent(Parcel p) {}

                internal static void Serialize(Parcel h, CallbackBase param)
                {
                    h.WriteInt(param.Id);
                    h.WriteInt(param.SeqId);
                    h.WriteBool(param.Once);
                }

                internal static void Deserialize(Parcel h, CallbackBase param)
                {
                    param.Id = h.ReadInt();
                    param.SeqId = h.ReadInt();
                    param.Once = h.ReadBool();
                }
            }

            public sealed class NotifyCB : CallbackBase
            {
                public NotifyCB(bool once = false) : base((int)DelegateId.NotifyCB, once)
                {
                }

                public delegate void Callback(string sender, string msg);
                public event Callback Received;

                internal override void OnReceivedEvent(Parcel parcel)
                {
                    string param1 = parcel.ReadString();
                    string param2 = parcel.ReadString();
                    Received?.Invoke(param1, param2);
                }

            }

            private enum DelegateId : int
            {
                NotifyCB = 1,
            }


            private enum MethodId : int
            {
                __Result = 0,
                __Callback = 1,
                Register = 2,
                Unregister = 3,
                Send = 4,

            }

            protected override void OnConnectedEvent(string endPoint, string portName, Port port)
            {
                _online = true;
                Connected?.Invoke(this, null);
            }

            protected override void OnDisconnectedEvent(string endPoint, string portName)
            {
                _online = false;
                Disconnected?.Invoke(this, null);
            }

            protected override void OnRejectedEvent(string endPoint, string portName)
            {
                Rejected?.Invoke(this, null);
            }

            private void ProcessReceivedEvent(Parcel parcel)
            {
                int id = parcel.ReadInt();
                int seqId = parcel.ReadInt();
                bool once = parcel.ReadBool();

                foreach (var i in _delegateList)
                {
                    if ((int)i.Id == id && i.SeqId == seqId)
                    {
                        i.OnReceivedEvent(parcel);
                        if (i.Once)
                            _delegateList.Remove(i);
                        break;
                    }
                }
            }

            protected override void OnReceivedEvent(string endPoint, string portName)
            {
                Parcel parcelReceived;

                try
                {
                    parcelReceived = new Parcel(CallbackPort);
                }
                catch (InvalidIOException)
                {
                    return;
                }

                using (parcelReceived)
                {
                    int cmd = parcelReceived.ReadInt();
                    if (cmd != (int)MethodId.__Callback)
                    {
                        return;
                    }

                    ProcessReceivedEvent(parcelReceived);
                }
            }

            private void ConsumeCommand(out Parcel parcel, Port port)
            {
                do
                {
                    Parcel p;

                    try
                    {
                        p = new Parcel(port);
                    }
                    catch (InvalidIOException)
                    {
                        parcel = null;
                        return;
                    }

                    int cmd = p.ReadInt();
                    if (cmd == (int)MethodId.__Result)
                    {
                        parcel = p;
                        return;
                    }

                    p.Dispose();
                    parcel = null;
                } while (true);
            }

            public Message(string appId) => _appId = appId;

            /// <summary>
            /// Connects to the stub app.
            /// </summary>
            /// <privilege>http://tizen.org/privilege/appmanager.launch</privilege>
            /// <privilege>http://tizen.org/privilege/datasharing</privilege>
            /// <exception cref="InvalidIDException">
            /// Thrown when the appid to connect is invalid.
            /// </exception>
            /// <exception cref="InvalidIOException">
            /// Thrown when internal I/O error happen.
            /// </exception>
            /// <exception cref="PermissionDeniedException">
            /// Thrown when the permission is denied.
            /// </exception>
            /// <remark> If you want to use this method, you must add privileges.</remark>
            public void Connect()
            {
                Connect(_appId, "Message");
            }

            /// <summary>
            /// Disconnects from the stub app.
            /// </summary>
            /// <exception cref="System.InvalidOperationException">
            /// Thrown when an internal IO error occurrs.
            /// </exception>
            public void Disconnect()
            {
                Port?.Disconnect();
            }

            /// <summary>
            /// Connects to the stub app synchronously.
            /// </summary>
            /// <privilege>http://tizen.org/privilege/appmanager.launch</privilege>
            /// <privilege>http://tizen.org/privilege/datasharing</privilege>
            /// <exception cref="InvalidIDException">
            /// Thrown when the appid to connect is invalid.
            /// </exception>
            /// <exception cref="InvalidIOException">
            /// Thrown when internal I/O error happen.
            /// </exception>
            /// <exception cref="PermissionDeniedException">
            /// Thrown when the permission is denied.
            /// </exception>
            /// <remark> If you want to use this method, you must add privileges.</remark>
            public void ConnectSync()
            {
                ConnectSync(_appId, "Message");
            }

            /// <summary>
            /// Disposes delegate objects in this interface
            /// </summary>
            /// <param name="tag">The tag string from delegate object</param>
            void DisposeCallback(string tag)
            {
                foreach (var i in _delegateList)
                {
                    if (i.Tag.Equals(tag))
                    {
                        _delegateList.Remove(i);
                        return;
                    }
                }
            }

            public int Register(string name, NotifyCB cb)
            {
                if (!_online)
                    throw new NotConnectedSocketException();

                using (Parcel p = new Parcel())
                {
                    ParcelHeader header = p.GetHeader();
                    header.SetTag(_tidlVersion);
                    p.WriteInt((int)MethodId.Register);
                    p.WriteString(name);
                    CallbackBase.Serialize(p, cb);

                    lock (_lock)
                    {
                        _delegateList.Add(cb);

                        // Send
                        p.Send(Port);

                    }

                    Parcel parcelReceived;
                    bool done = false;
                    do
                    {
                        lock (_lock)
                        {
                            // Receive
                            ConsumeCommand(out parcelReceived, Port);
                        }

                        if (parcelReceived == null)
                        {
                            throw new InvalidProtocolException();
                        }

                        ParcelHeader headerReceived = parcelReceived.GetHeader();
                        if (!string.IsNullOrEmpty(headerReceived.GetTag()))
                        {
                            if (headerReceived.GetSequenceNumber() != header.GetSequenceNumber())
                            {
                                parcelReceived.Dispose();
                                parcelReceived = null;
                            }
                        }

                        if (parcelReceived != null)
                            done = true;
                    }
                    while (!done);

                    int ret = parcelReceived.ReadInt();
                    parcelReceived.Dispose();
                    return ret;
                }
            }

            public void Unregister()
            {
                if (!_online)
                    throw new NotConnectedSocketException();

                using (Parcel p = new Parcel())
                {
                    ParcelHeader header = p.GetHeader();
                    header.SetTag(_tidlVersion);
                    p.WriteInt((int)MethodId.Unregister);

                    lock (_lock)
                    {
                        // Send
                        p.Send(Port);

                    }
                }
            }

            public int Send(string msg)
            {
                if (!_online)
                    throw new NotConnectedSocketException();

                using (Parcel p = new Parcel())
                {
                    ParcelHeader header = p.GetHeader();
                    header.SetTag(_tidlVersion);
                    p.WriteInt((int)MethodId.Send);
                    p.WriteString(msg);

                    lock (_lock)
                    {
                        // Send
                        p.Send(Port);

                    }

                    Parcel parcelReceived;
                    bool done = false;
                    do
                    {
                        lock (_lock)
                        {
                            // Receive
                            ConsumeCommand(out parcelReceived, Port);
                        }

                        if (parcelReceived == null)
                        {
                            throw new InvalidProtocolException();
                        }

                        ParcelHeader headerReceived = parcelReceived.GetHeader();
                        if (!string.IsNullOrEmpty(headerReceived.GetTag()))
                        {
                            if (headerReceived.GetSequenceNumber() != header.GetSequenceNumber())
                            {
                                parcelReceived.Dispose();
                                parcelReceived = null;
                            }
                        }

                        if (parcelReceived != null)
                            done = true;
                    }
                    while (!done);

                    int ret = parcelReceived.ReadInt();
                    parcelReceived.Dispose();
                    return ret;
                }
            }

        }

    }
}
}
```

Reply:
Thanks for this suggestion.
Unfortunately, the TIDL interface needs to be changed somewhat in order to apply your proposal.
As an alternative, I have partially refactored the proxy app main.
The app can update messages by creating a MyMessage interface without global variables.
How about this? Could you give your opinion?